rlw/upgrade.ps1

```diff
@@ -0,0 +1,167 @@
+      Stop-Process -Name 'rightlink' -Force
+      Exit 1
+    }
+
+    # Report to audit entry that RightLink ugpraded.
+    $instanceHref = & ${RIGHTLINK_DIR}\rsc.exe --rl10 --x1 ':has(.rel:val(\"\"\"self\"\"\")).href' `
+                    cm15 index_instance_session /api/sessions/instance 2> $null
+    if ($instanceHref) {
+      $auditEntryHref = & ${RIGHTLINK_DIR}\rsc.exe --rl10 --xh 'location' cm15 create /api/audit_entries `
+                        """audit_entry[auditee_href]=${instanceHref}""" `
+                        """audit_entry[detail]=RightLink updated to '${newInstalledVersion}'""" `
+                        """audit_entry[summary]=RightLink updated""" 2> $null
+      if ($auditEntryHref) {
+        Write-Output """Audit entry created at ${auditEntryHref}"""
+      } else {
+        Write-Output 'Failed to create audit entry'
+      }
+    } else {
+      Write-Output 'Unable to obtain instance HREF for audit entries'
+    }
+  }
+}
+
+$RIGHTLINK_DIR = 'C:\Program Files\RightScale\RightLink'
+
+if (!$env:UPGRADES_FILE_LOCATION) {
+  $env:UPGRADES_FILE_LOCATION = 'https://rightlink.rightscale.com/rightlink/upgrades'
+}
+
+# Query RightLink info
+$json = & "${RIGHTLINK_DIR}\rsc.exe" rl10 index /rll/proc
+
+# Determine current version of rightlink
+$currentVersion = Write-Output $json | & "${RIGHTLINK_DIR}\rsc.exe" --x1 .version json 2> $null
+
+if (!$currentVersion) {
+  Write-Output 'Cannot determine current version of RightLink'
+  Exit 1
+}
+
+# Fetch information about what we should become. The "upgrades" file consists of lines formatted
+# as "currentVersion:desiredVersion". If the "upgrades" file does not exist,
+# or if the current version is not in the file, no upgrade is done.
+$upgradesFile = "${PSScriptRoot}\upgrades"
+(New-Object System.Net.WebClient).DownloadFile($env:UPGRADES_FILE_LOCATION, $upgradesFile)
+$desiredVersion = Get-Content $upgradesFile | Select-String "^\s*${currentVersion}\s*:\s*(\S+)\s*$" |
+                  % { $_ -replace "^\s*${currentVersion}\s*:\s*", '' }
+
+if (!$desiredVersion) {
+  Write-Output 'Cannot determine latest version from upgrade file'
+  Write-Output "Tried to match /^${currentVersion}:/ in ${env:UPGRADES_FILE_LOCATION}"
+  Exit 0
+}
+
+if ($desiredVersion -eq $currentVersion) {
+  Write-Output "RightLink is already up-to-date (current = ${currentVersion})"
+  Exit 0
+}
+
+Write-Output 'RightLink needs update:'
+Write-Output "  from current = ${currentVersion}"
+Write-Output "  to   desired = ${desiredVersion}"
+
+Write-Output "Downloading RightLink version '${desiredVersion}'"
+
+# Download new version
+$TMP_DIR = 'C:\Windows\Temp\Upgrade'
+$RIGHTLINK_URL = "https://rightlink.rightscale.com/rll/${desiredVersion}/rightlink.zip"
+
+# If temp directory doesn't exist then create it
+if (!(Test-Path -Path $TMP_DIR)) {
+  New-Item -Path $TMP_DIR -Type Directory | Out-Null
+}
+
```

Comment:
Are the following few conditional `Remove-Item`s to clear out the temp directory before putting stuff in it? If so, there should probably be a comment here explaining that.